pkg/distributor/distributor_test.go

```diff
@@ -1467,6 +1467,244 @@ func TestDistributor_Push_HistogramValidation(t *testing.T) {
+			req: makeWriteRequestWith(
+				makeTimeseries(labels, append(makeSamples(ts1, v1), makeSamples(ts1, v2)...), makeHistograms(ts3, generateTestHistogram(0)), nil),
+				makeTimeseries(labels, makeSamples(ts1, v3), append(makeHistograms(ts2, generateTestHistogram(1)), makeHistograms(ts2, generateTestHistogram(2))...), nil),
+				makeTimeseries(labels, makeSamples(ts1, v4), append(makeHistograms(ts2, generateTestHistogram(3)), makeHistograms(ts3, generateTestHistogram(4))...), nil),
+			),
+			expectedSamples: []mimirpb.PreallocTimeseries{
+				makeTimeseries(labels, makeSamples(ts1, v1), makeHistograms(ts3, generateTestHistogram(0)), nil),
+				makeTimeseries(labels, makeSamples(ts1, v3), makeHistograms(ts2, generateTestHistogram(1)), nil),
+				makeTimeseries(labels, makeSamples(ts1, v4), append(makeHistograms(ts2, generateTestHistogram(3)), makeHistograms(ts3, generateTestHistogram(4))...), nil),
+			},
+			expectedErrors: []error{
+				fmt.Errorf("samples with duplicated timestamps have been discarded, discarded samples: %d series: '%.200s' (err-mimir-sample-duplicate-timestamp)", 1, metricName),
+				fmt.Errorf("samples with duplicated timestamps have been discarded, discarded samples: %d series: '%.200s' (err-mimir-sample-duplicate-timestamp)", 1, metricName),
+				nil,
+			},
+			expectedMetrics: `
+				# HELP cortex_discarded_samples_total The total number of samples that were discarded.
+				# TYPE cortex_discarded_samples_total counter
+				cortex_discarded_samples_total{group="test-group",reason="sample_duplicate_timestamp",user="user"} 2
+			`,
+		},
+	}
+
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			limits := prepareDefaultLimits()
+			ds, _, regs, _ := prepare(t, prepConfig{
+				limits:          limits,
+				numDistributors: 1,
+			})
+
+			// Pre-condition check.
+			require.Len(t, ds, 1)
+			require.Len(t, regs, 1)
+
+			now := mtime.Now()
+			for i, ts := range tc.req.Timeseries {
+				shouldRemove, err := ds[0].validateSeries(now, &ts, "user", "test-group", true, true, 0, 0)
+				require.False(t, shouldRemove)
+				if len(tc.expectedErrors) == 0 {
+					require.NoError(t, err)
+				} else {
+					if tc.expectedErrors[i] == nil {
+						require.NoError(t, err)
+					} else {
+						require.Error(t, err)
+						require.Equal(t, tc.expectedErrors[i], err)
+					}
+				}
+			}
+
+			assert.Equal(t, tc.expectedSamples, tc.req.Timeseries)
+			assert.NoError(t, testutil.GatherAndCompare(regs[0], strings.NewReader(tc.expectedMetrics), "cortex_discarded_samples_total"))
+		})
+	}
+}
+
+func BenchmarkDistributor_SampleDuplicateTimestamp(b *testing.B) {
+	const (
+		metricName              = "series"
+		testSize                = 80_000
+		numberOfDifferentValues = 40_000
+	)
+	labels := []string{labels.MetricName, metricName, "job", "job", "service", "service"}
+
+	now := mtime.Now()
+	timestamp := now.UnixMilli()
+
+	testCases := map[string]struct {
```

Comment:
You tested with extreme edge cases, but I would like to also see a benchmark for the happy path
- 1 sample per series, no duplicated samples
- 2 samples per series, no duplicated samples

Then, run this benchmark both against `main` and this PR (run each benchmark at least 6 times). And then compare the results with `benchcmp` (please post the updated benchmark in the PR description).

Reply:
I have added the following benchmarks:
- 1 sample per series, no duplicated samples
- 1 histogram per series, no duplicated samples
- 1 sample and 1 histogram per series, no duplicated samples
- 2 samples per series, no duplicated samples
- 2 histograms per series, no duplicated samples
- 2 samples and 2 histograms per series, no duplicated samples